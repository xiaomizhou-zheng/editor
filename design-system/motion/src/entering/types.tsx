/**
 * Direction the motion is going.
 */
export type Transition = 'entering' | 'exiting';

/**
 * Direction an element enters from or leaves towards
 */
export type Direction = 'top' | 'right' | 'bottom' | 'left';

/**
 * The distance the element moves in a direction-based animation.
 * A `proportional` distance is based on the size of the element.
 * A `constant` distance will always move the same amount, regardless of size.
 */
export type Distance = 'constant' | 'proportional';

/**
 * Whether an element will fade on enter, on exit or both
 */
export type Fade = 'none' | 'in' | 'out' | 'inout';

/**
 * Common props all entering motions should make available for consumers.
 */
export interface MotionProps<TProps extends {}> {
  /**
   * Duration in `ms`.
   * How long the motion will take.
   */
  duration?: number;

  /**
   * Will callback when the motion has finished in the particular direction.
   * If it finished entering direction will be `entering`.
   * And vice versa for `exiting`.
   */
  onFinish?: (state: Transition) => void;

  /**
   * Children as `function`.
   * Will be passed `props` for you to hook up.
   * The `state` arg can be used to know if the motion is `entering` or `exiting`.
   */
  children: (opts: TProps, state: Transition) => React.ReactNode;
}
