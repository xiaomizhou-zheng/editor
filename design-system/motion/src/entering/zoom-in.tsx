import React from 'react';

import { CSSObject } from '@emotion/react';

import KeyframesMotion, { KeyframesMotionProps } from './keyframes-motion';

export const zoomInAnimation = (): CSSObject => ({
  '0%': {
    opacity: 0,
    transform: 'scale(0.5)',
  },
  '50%': {
    opacity: 1,
  },
  '75%': {
    transform: 'scale(1.25)',
  },
  '100%': {
    transform: 'scale(1)',
  },
});

export const shrinkOutAnimation = (): CSSObject => ({
  to: {
    opacity: 0,
    transform: 'scale(0.75)',
  },
});

/**
 * __ZoomIn__
 *
 * Will over zoom an element into position.
 *
 * - [Examples](https://atlaskit.atlassian.com/packages/design-system/motion/docs/entering-motions)
 */
const ZoomIn: React.FC<KeyframesMotionProps> = ({
  children,
  duration = 125,
  isPaused,
  onFinish,
}: KeyframesMotionProps) => {
  return (
    <KeyframesMotion
      duration={duration}
      enteringAnimation={zoomInAnimation()}
      exitingAnimation={shrinkOutAnimation()}
      animationTimingFunction={() => 'ease-in-out'}
      isPaused={isPaused}
      onFinish={onFinish}
    >
      {children}
    </KeyframesMotion>
  );
};

export default ZoomIn;
