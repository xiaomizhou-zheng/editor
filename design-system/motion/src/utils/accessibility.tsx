import { useEffect, useState } from 'react';

import { bind } from 'bind-event-listener';

const isMatchMediaAvailable = (): boolean =>
  typeof window !== 'undefined' && 'matchMedia' in window;

/**
 * Use for any programatic motions needed at runtime.
 * Will return `true` if the current user prefers reduced motion.
 * This is generally set through OS preferences/settings.
 */
export const isReducedMotion = (): boolean => {
  if (!isMatchMediaAvailable()) {
    return false;
  }

  const { matches } = window.matchMedia('(prefers-reduced-motion: reduce)');
  return matches;
};

/**
 * A React hook version of {@link isReducedMotion}.
 * Useful for React components that need to re-render if the user's motion
 * preference changes at runtime.
 */
export const useIsReducedMotion = (): boolean => {
  const [prefersReducedMotion, setPrefersReducedMotion] = useState(
    isReducedMotion,
  );

  useEffect(() => {
    if (!isMatchMediaAvailable()) {
      return;
    }

    const mediaQueryList = window.matchMedia(
      '(prefers-reduced-motion: reduce)',
    );

    const onChange = (event: MediaQueryListEvent) =>
      setPrefersReducedMotion(event.matches);

    return bind(mediaQueryList, {
      type: 'change',
      listener: onChange,
    });
  }, []);

  return prefersReducedMotion;
};

/**
 * Use for any CSS based motion (animation or transition).
 * Always put at the end of your declaration for correct use of the cascade.
 * Reduced motion preference is generally set through OS preferences/settings.
 */
export const prefersReducedMotion = () => ({
  '@media (prefers-reduced-motion: reduce)': {
    animation: 'none',
    transition: 'none',
  },
});
