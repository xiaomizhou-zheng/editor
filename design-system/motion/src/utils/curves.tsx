export type AnimationCurve =
  | 'cubic-bezier(0.15,1,0.3,1)'
  | 'cubic-bezier(0.2,0,0,1)'
  | 'cubic-bezier(0.8,0,0,0.8)';

export const easeInOut: AnimationCurve = 'cubic-bezier(0.15,1,0.3,1)';

export const easeOut: AnimationCurve = 'cubic-bezier(0.2,0,0,1)';

export const easeIn: AnimationCurve = 'cubic-bezier(0.8,0,0,0.8)';
