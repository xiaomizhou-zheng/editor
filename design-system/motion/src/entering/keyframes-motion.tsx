import React, { Ref, useEffect, useState } from 'react';

import { ClassNames, CSSObject, keyframes } from '@emotion/react';

import { prefersReducedMotion } from '../utils/accessibility';
import { largeDurationMs } from '../utils/durations';
import { useSetTimeout } from '../utils/timer-hooks';

import { useExitingPersistence } from './exiting-persistence';
import { useStaggeredEntrance } from './staggered-entrance';
import { MotionProps, Transition } from './types';

/**
 * These are props that motions should use as their external props for consumers.
 * See [FadeIn](packages/helpers/motion/src/entering/fade-in.tsx) for an example usage.
 */
export interface KeyframesMotionProps
  extends MotionProps<{ className: string; ref: Ref<any> }> {
  /**
   * Can be used to pause the animation before it has finished.
   */
  isPaused?: boolean;
}

interface InternalKeyframesMotionProps extends KeyframesMotionProps {
  /**
   * Timing function to be used with the animation.
   * Receives the `state` and expects a `string` return value.
   * Useful if you want a different curve when entering vs. exiting.
   */
  animationTimingFunction: (state: Transition) => string;

  /**
   * CSS keyframes for the entering animation.
   */
  enteringAnimation: CSSObject;

  /**
   * CSS keyframes for the exiting animation.
   */
  exitingAnimation?: CSSObject;

  /**
   * Duration in `ms`.
   * How long the motion will take.
   */
  duration: number;
}

/**
 * Used to multiply the initial duration for exiting motions.
 */
const EXITING_MOTION_MULTIPLIER = 0.5;

/**
 * This is the base INTERNAL component used for all other entering motions.
 * This does not need Javascript to execute on the client so it will run immediately
 * for any SSR rendered React apps before the JS has executed.
 */
const EnteringMotion: React.FC<InternalKeyframesMotionProps> = ({
  children,
  animationTimingFunction,
  enteringAnimation,
  exitingAnimation,
  isPaused,
  onFinish: onFinishMotion,
  duration = largeDurationMs,
}: InternalKeyframesMotionProps) => {
  const staggered = useStaggeredEntrance();
  const {
    isExiting,
    onFinish: onExitFinished,
    appear,
  } = useExitingPersistence();
  const setTimeout = useSetTimeout();
  const paused = isPaused || !staggered.isReady;
  const delay = isExiting ? 0 : staggered.delay;
  const state = isExiting ? 'exiting' : 'entering';
  const [hasAnimationStyles, setHasAnimationStyles] = useState(appear);

  useEffect(() => {
    // Tracking this to prevent changing state on an unmounted component
    let isCancelled = false;

    if (paused) {
      return;
    }

    // On initial mount if elements aren't set to animate on appear, we return early and callback
    if (!appear) {
      onFinishMotion && onFinishMotion(state);
      return;
    }

    // Elements may need animation styles back after initial mount (they could animate out)
    setHasAnimationStyles(true);

    setTimeout(
      () => {
        if (state === 'exiting') {
          onExitFinished && onExitFinished();
        }
        if (!isCancelled) {
          setHasAnimationStyles(false);
        }
        onFinishMotion && onFinishMotion(state);
      },
      isExiting ? duration * EXITING_MOTION_MULTIPLIER : duration + delay,
    );

    return () => {
      isCancelled = true;
    };
    // We ignore this for onFinishMotion as consumers could potentially inline the function
    // which would then trigger this effect every re-render.
    // We want to make it easier for consumers so we go down this path unfortunately.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [onExitFinished, state, isExiting, duration, delay, paused, setTimeout]);

  return (
    <ClassNames>
      {({ css }) =>
        children(
          {
            ref: staggered.ref,
            className: hasAnimationStyles
              ? css({
                  ...prefersReducedMotion(),
                  animationDelay: `${delay}ms`,
                  animationDuration: `${
                    isExiting ? duration * EXITING_MOTION_MULTIPLIER : duration
                  }ms`,
                  animationFillMode: isExiting ? 'forwards' : 'backwards',
                  animationName: `${keyframes(
                    isExiting
                      ? exitingAnimation || enteringAnimation
                      : enteringAnimation,
                  )}`,
                  animationPlayState: paused ? 'paused' : 'running',
                  animationTimingFunction: animationTimingFunction(state),
                })
              : '',
          },
          state,
        )
      }
    </ClassNames>
  );
};

export default EnteringMotion;
