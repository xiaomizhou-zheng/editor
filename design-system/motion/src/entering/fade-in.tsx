import React from 'react';

import { CSSObject } from '@emotion/react';

import { easeInOut } from '../utils/curves';
import { largeDurationMs } from '../utils/durations';

import KeyframesMotion, { KeyframesMotionProps } from './keyframes-motion';
import type { Direction, Distance } from './types';

const directionMotions: Record<Distance, Record<Direction, string>> = {
  proportional: {
    bottom: 'translate3d(0, calc(5% + 4px), 0)',
    left: 'translate3d(calc(-5% - 4px), 0, 0)',
    right: 'translate3d(calc(5% + 4px), 0, 0)',
    top: 'translate3d(0, calc(-5% - 4px), 0)',
  },
  constant: {
    bottom: 'translate3d(0, 4px, 0)',
    left: 'translate3d(-4px, 0, 0)',
    right: 'translate3d(4px, 0, 0)',
    top: 'translate3d(0, -4px, 0)',
  },
};

const invertedDirection = {
  top: 'bottom',
  bottom: 'top',
  left: 'right',
  right: 'left',
} as const;

export const fadeInAnimation = (
  direction?: Direction,
  distance: Distance = 'proportional',
): CSSObject => {
  return {
    from: {
      opacity: 0,
      ...(direction !== undefined && {
        transform: directionMotions[distance][direction],
      }),
    },
    '50%': {
      opacity: 1,
    },
    to: {
      transform: direction !== undefined ? 'none' : undefined,
    },
  };
};

export const fadeOutAnimation = (
  direction?: Direction,
  distance: Distance = 'proportional',
): CSSObject => ({
  from: {
    opacity: 1,
    transform: direction !== undefined ? 'translate3d(0, 0, 0)' : undefined,
  },
  to: {
    opacity: 0,
    ...(direction !== undefined && {
      transform: directionMotions[distance][direction],
    }),
  },
});

/**
 * Props for controlling the behavior of the FadeIn animation
 */
export interface FadeKeyframesMotionProps extends KeyframesMotionProps {
  /**
   * The direction the element will enter from using a slide animation. If undefined, no slide will be applied.
   */
  entranceDirection?: Direction;
  /**
   * The direction the element will exit to using a slide animation.
   * If undefined but `entranceDirection` is set, it will exit to the opposite side by default.
   */
  exitDirection?: Direction;
  /**
   * The distance the element moves in a direction-based animation.
   * A `proportional` distance is based on the size of the element.
   * A `constant` distance will always move the same amount, regardless of size.
   */
  distance?: Distance;
}

/**
 * __FadeIn__
 *
 * Useful for fading in one or more elements.
 *
 * - [Examples](https://atlaskit.atlassian.com/packages/design-system/motion/docs/entering-motions)
 */
const FadeIn: React.FC<FadeKeyframesMotionProps> = ({
  children,
  duration = largeDurationMs,
  entranceDirection,
  exitDirection,
  distance = 'proportional',
  onFinish,
  isPaused,
}: FadeKeyframesMotionProps) => {
  const invertedEntranceDirection =
    entranceDirection && invertedDirection[entranceDirection];

  return (
    <KeyframesMotion
      duration={duration}
      enteringAnimation={fadeInAnimation(entranceDirection, distance)}
      exitingAnimation={fadeOutAnimation(
        exitDirection || invertedEntranceDirection,
        distance,
      )}
      animationTimingFunction={() => easeInOut}
      onFinish={onFinish}
      isPaused={isPaused}
    >
      {children}
    </KeyframesMotion>
  );
};

export default FadeIn;
