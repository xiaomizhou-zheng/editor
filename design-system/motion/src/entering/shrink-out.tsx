import React from 'react';

import { easeIn } from '../utils/curves';
import { smallDurationMs } from '../utils/durations';
import { useRequestAnimationFrame, useSetTimeout } from '../utils/timer-hooks';
import { useElementRef } from '../utils/use-element-ref';
import { useLayoutEffect } from '../utils/use-layout-effect';

import { useExitingPersistence } from './exiting-persistence';
import { MotionProps } from './types';

export interface ShrinkOutProps extends MotionProps<{ ref: React.Ref<any> }> {}

/**
 * __ShrinkOut__
 *
 * Will shrink an element down to nothing when exiting.
 * Works best with flex children as collapsing margins can come with undesired behaviour.
 *
 * - [Examples](https://atlaskit.atlassian.com/packages/design-system/motion/docs/entering-motions)
 */
const ShrinkOut: React.FC<ShrinkOutProps> = ({
  children,
  duration = smallDurationMs,
  onFinish,
}: ShrinkOutProps): any => {
  const [element, setElementRef] = useElementRef();
  const exiting = useExitingPersistence();
  const requestAnimationFrame = useRequestAnimationFrame();
  const setTimeout = useSetTimeout();
  const direction = exiting.isExiting ? 'exiting' : 'entering';

  useLayoutEffect(() => {
    if (exiting.isExiting && element) {
      const newStyles: React.CSSProperties = {
        // We fix both width and height because when changing box sizing to border-box.
        width: `${element.offsetWidth}px`,
        height: `${element.offsetHeight}px`,
        boxSizing: 'border-box',
        willChange: 'width,margin',
      };
      Object.assign(element.style, newStyles);

      requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          const newStyles: React.CSSProperties = {
            width: '0px',
            margin: '0px', // We animate margin down to zero so it doesn't take any space.
            transitionTimingFunction: easeIn,
            transitionDuration: `${duration}ms`,
            transitionProperty: 'width,margin',
          };
          Object.assign(element.style, newStyles);

          setTimeout(() => {
            exiting.onFinish && exiting.onFinish();
            onFinish && onFinish('exiting');
          }, duration);
        });
      });
    }
  });

  return children({ ref: setElementRef }, direction);
};

export default ShrinkOut;
