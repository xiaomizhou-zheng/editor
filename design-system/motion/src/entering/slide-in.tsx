import React from 'react';

import { CSSObject } from '@emotion/react';

import { AnimationCurve, easeIn, easeOut } from '../utils/curves';
import { mediumDurationMs } from '../utils/durations';

import KeyframesMotion, { KeyframesMotionProps } from './keyframes-motion';
import { Direction, Fade, Transition } from './types';

export const slideInAnimation = (
  from: Direction,
  state: Transition,
  fade: Fade,
): CSSObject => {
  const initial = state === 'entering' ? '0%' : '100%';
  const end = state === 'entering' ? '100%' : '0%';
  const fromMap = {
    top: 'translate3d(0, -100%, 0)',
    right: 'translate3d(100%, 0, 0)',
    bottom: 'translate3d(0, 100%, 0)',
    left: 'translate3d(-100%, 0, 0)',
  };

  const animateOpacity =
    fade === 'inout' ||
    (fade === 'in' && state === 'entering') ||
    (fade === 'out' && state === 'exiting');

  return {
    [initial]: {
      transform: fromMap[from],
      opacity: animateOpacity ? 0 : 1,
    },
    [end]: {
      transform: 'none',
      opacity: 1,
    },
  };
};

export interface SlideInProps extends KeyframesMotionProps {
  /**
   * Direction the element will slide in from.
   * E.g. `"right"` will slide in from the right to the left.
   *
   * If `exitTo` is not set, exiting will reverse this motion.
   * E.g. if `enterFrom: "right"`, will slide out to the right.
   */
  enterFrom: Direction;

  /**
   * Direction the element will slide out towards.
   * E.g. `"right"` will slide out to the right.
   *
   * If this is not set, exiting will reverse the entrance motion.
   * E.g. if `enterFrom: "right"`, will slide out to the right.
   */
  exitTo?: Direction;

  /**
   * Whether an element will fade on enter, on exit or both.
   *
   * `'none'` is the default and will cause the element to not fade,
   * `'in'` will cause the element to fade on enter,
   * `'out'` will cause the element to fade on exit,
   * `'inout'` will cause the element to fade on both
   */
  fade?: Fade;

  /**
   * A custom function used to override the default animation timings.
   * Returns which animation curve to use depending on the current transitioning state.
   */
  animationTimingFunction?: (state: Transition) => AnimationCurve;
}

/**
 * __SlideIn__
 *
 * Will slide an element into position, generally used for things that appear from outside of the viewport into view.
 *
 * - [Examples](https://atlaskit.atlassian.com/packages/design-system/motion/docs/entering-motions)
 */
const SlideIn: React.FC<SlideInProps> = ({
  children,
  enterFrom,
  exitTo,
  fade = 'none',
  duration = mediumDurationMs,
  isPaused,
  onFinish,
  animationTimingFunction = (state) =>
    state === 'entering' ? easeOut : easeIn,
}: SlideInProps) => {
  return (
    <KeyframesMotion
      duration={duration}
      enteringAnimation={slideInAnimation(enterFrom, 'entering', fade)}
      exitingAnimation={slideInAnimation(exitTo || enterFrom, 'exiting', fade)}
      animationTimingFunction={animationTimingFunction}
      isPaused={isPaused}
      onFinish={onFinish}
    >
      {children}
    </KeyframesMotion>
  );
};

export default SlideIn;
