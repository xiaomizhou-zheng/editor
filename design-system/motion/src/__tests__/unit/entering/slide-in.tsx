import React from 'react';

import { render } from '@testing-library/react';

import ExitingPersistence from '../../../entering/exiting-persistence';
import SlideIn, { slideInAnimation } from '../../../entering/slide-in';
import { Direction, Fade, Transition } from '../../../entering/types';
import { easeIn, easeOut } from '../../../utils/curves';
import { mediumDurationMs } from '../../../utils/durations';

jest.mock('../../../utils/accessibility');

describe('<SlideIn />', () => {
  it('should default to medium duration', () => {
    const { getByTestId } = render(
      <SlideIn enterFrom="left">
        {(props) => <div data-testid="target" {...props} />}
      </SlideIn>,
    );

    expect(getByTestId('target')).toHaveStyleDeclaration(
      'animation-duration',
      `${mediumDurationMs}ms`,
    );
  });

  it('should override default duration', () => {
    const { getByTestId } = render(
      <SlideIn duration={123} enterFrom="left">
        {(props) => <div data-testid="target" {...props} />}
      </SlideIn>,
    );

    expect(getByTestId('target')).toHaveStyleDeclaration(
      'animation-duration',
      '123ms',
    );
  });

  it('should slide in easing out', () => {
    const { getByTestId } = render(
      <SlideIn enterFrom="left">
        {(props) => <div data-testid="target" {...props} />}
      </SlideIn>,
    );

    expect(getByTestId('target')).toHaveStyleDeclaration(
      'animation-timing-function',
      easeOut,
    );
  });

  it('should slide out easing in', () => {
    const { getByTestId, rerender } = render(
      <ExitingPersistence>
        <SlideIn enterFrom="left">
          {(props) => <div data-testid="target" {...props} />}
        </SlideIn>
      </ExitingPersistence>,
    );

    rerender(<ExitingPersistence>{false}</ExitingPersistence>);

    expect(getByTestId('target')).toHaveStyleDeclaration(
      'animation-timing-function',
      easeIn,
    );
  });

  ['entering', 'exiting'].forEach((state) => {
    ['top', 'right', 'bottom', 'left'].forEach((from) => {
      ['none', 'in', 'out', 'inout'].forEach((fade) => {
        it(`should animate ${state} from ${from} with fade ${fade}`, () => {
          expect(
            slideInAnimation(
              from as Direction,
              state as Transition,
              fade as Fade,
            ),
          ).toMatchSnapshot();
        });
      });
    });
  });
});
