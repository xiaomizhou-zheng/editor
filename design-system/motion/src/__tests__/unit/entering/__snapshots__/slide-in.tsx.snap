// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SlideIn /> should animate entering from bottom with fade in 1`] = `
Object {
  "0%": Object {
    "opacity": 0,
    "transform": "translate3d(0, 100%, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from bottom with fade inout 1`] = `
Object {
  "0%": Object {
    "opacity": 0,
    "transform": "translate3d(0, 100%, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from bottom with fade none 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "translate3d(0, 100%, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from bottom with fade out 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "translate3d(0, 100%, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from left with fade in 1`] = `
Object {
  "0%": Object {
    "opacity": 0,
    "transform": "translate3d(-100%, 0, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from left with fade inout 1`] = `
Object {
  "0%": Object {
    "opacity": 0,
    "transform": "translate3d(-100%, 0, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from left with fade none 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "translate3d(-100%, 0, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from left with fade out 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "translate3d(-100%, 0, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from right with fade in 1`] = `
Object {
  "0%": Object {
    "opacity": 0,
    "transform": "translate3d(100%, 0, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from right with fade inout 1`] = `
Object {
  "0%": Object {
    "opacity": 0,
    "transform": "translate3d(100%, 0, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from right with fade none 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "translate3d(100%, 0, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from right with fade out 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "translate3d(100%, 0, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from top with fade in 1`] = `
Object {
  "0%": Object {
    "opacity": 0,
    "transform": "translate3d(0, -100%, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from top with fade inout 1`] = `
Object {
  "0%": Object {
    "opacity": 0,
    "transform": "translate3d(0, -100%, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from top with fade none 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "translate3d(0, -100%, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate entering from top with fade out 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "translate3d(0, -100%, 0)",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "none",
  },
}
`;

exports[`<SlideIn /> should animate exiting from bottom with fade in 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "translate3d(0, 100%, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from bottom with fade inout 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 0,
    "transform": "translate3d(0, 100%, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from bottom with fade none 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "translate3d(0, 100%, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from bottom with fade out 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 0,
    "transform": "translate3d(0, 100%, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from left with fade in 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "translate3d(-100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from left with fade inout 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 0,
    "transform": "translate3d(-100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from left with fade none 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "translate3d(-100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from left with fade out 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 0,
    "transform": "translate3d(-100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from right with fade in 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "translate3d(100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from right with fade inout 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 0,
    "transform": "translate3d(100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from right with fade none 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "translate3d(100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from right with fade out 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 0,
    "transform": "translate3d(100%, 0, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from top with fade in 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "translate3d(0, -100%, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from top with fade inout 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 0,
    "transform": "translate3d(0, -100%, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from top with fade none 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 1,
    "transform": "translate3d(0, -100%, 0)",
  },
}
`;

exports[`<SlideIn /> should animate exiting from top with fade out 1`] = `
Object {
  "0%": Object {
    "opacity": 1,
    "transform": "none",
  },
  "100%": Object {
    "opacity": 0,
    "transform": "translate3d(0, -100%, 0)",
  },
}
`;
